tests/gradients/test_finite_difference.py

```diff
@@ -281,13 +239,13 @@ def test_independent_parameters(self):
         dev = qml.device("default.qubit", wires=2)
 
         with qml.tape.QuantumTape() as tape1:
-            qml.RX(1, wires=[0])
-            qml.RX(1, wires=[1])
+            qml.RX(1.0, wires=[0])

```

Comment:
How come this has changed?

Reply:
`finite_diff` uses `generate_shifted_tapes`, which now has been update to cast shifted parameters to the original dtype. For integers, this goes wrong, because `int(1+1e-7)=1`, so that no shift happens.

Reply:
@dwierichs I wonder if, in some cases, users might view this as a bug or not 🤔

Actually this does feel a bit like a bug for me. If we are shifting, and the shift is a float, then the new parameter should also be a float --- even if the original was an int.

The final cast back to `int()` definitely feels wrong 🤔

Reply:
Mhm. This is really tricky, because it would break custom data type handling, wouldn't it? :thinking: 

What would you suggest here? We could:
- Make an exception for `int`s and similar dtypes (like `tf.int32`) or maybe
- Not change the behaviour but raise a warning.

Reply:
Let me think about it 🤔

Reply:
Okay, after thinking about it some more --- and separate to this feeling like a bug --- this definitely feels like a breaking change! In that people with existing code using PL will suddenly have code with `RX(1, wires=0)` that _used_ to work, but will now silently give incorrect gradients.

So my main recommendation is that we should either:

- Ensure that `shift(int parameter, float shift) -> float shifted_parameter`, or
- Ensure that `shift(int parameter, float shift)` raises an error.

My preference is definitely for the first one, since this is how Autograd and NumPy works!

Although there is an option to support _both_, and just leave the type casting up to the framework. E.g., NumPy would allow int + float -> float, whereas TF would raise an error instructing the user to ensure the types are compatible

Reply:
While reading your last comment, I was already sure that I'd do the first, and simply always cast to `float` s the way you suggested above also.
But the last paragraph is intriguing! :exploding_head: Wouldn't it be the "most compatible" (and for users of the respective framework the most convenient) approach to have the framework to the casting? :)

Reply:
See above, I went for casting to `float` in all cases in which `new_params[idx].dtype` is not defined.